server/context.ts

```diff
@@ -1,28 +1,209 @@
 import { Session } from "@prisma/client";
 import { createServerClient } from "@/server/utils/supabase/supabase-server-client";
 import { jwtDecode } from "jwt-decode";
 import { prisma } from "./utils/prisma/prisma-client";
 import {
-  getClientCountryCode,
   getClientIp,
   getIpInfo,
 } from "./utils/ip/ip.utils";
 
+// Node.js equivalent of browser's atob function
+function nodeDecode(base64String: string): string {
+  return Buffer.from(base64String, 'base64').toString('utf-8');
+}
+
 export async function createContext({ req }: { req: Request }) {
+
   const authHeader = req.headers.get("authorization");
+  const customAuthHeader = req.headers.get("x-custom-auth");
   const clientId = req.headers.get("x-client-id");
   const applicationId = req.headers.get("x-application-id") || "";
 
-  if (!authHeader || !clientId) {
+  if ((!authHeader && !customAuthHeader) || !clientId) {
+    console.log("Missing required headers for authentication");
     return createEmptyContext();
   }
 
+  const userAgent = req.headers.get("user-agent") || "";
+  const deviceNonce = req.headers.get("x-device-nonce") || "";
+  let ip = getClientIp(req);
+
+  if (process.env.NODE_ENV === "development") {
+    const ipInfo = await getIpInfo();
+    if (ipInfo) {
+      ({ ip } = ipInfo);
+    }
+  }
+
+  // Prioritize the custom auth token if present
+  if (customAuthHeader) {
+    console.log("Processing custom auth token");
+    try {
+      // Parse the base64 encoded JSON token using Node.js Buffer
+      let decodedTokenString;
+      try {
+        decodedTokenString = nodeDecode(customAuthHeader);
+        console.log("Decoded token string length:", decodedTokenString.length);
+      } catch (decodeError) {
+        console.error("Failed to decode custom auth token:", decodeError);
+        console.log("Raw token (first 50 chars):", customAuthHeader.substring(0, 50));
+        return createEmptyContext();
+      }
+      
+      let decodedToken;
+      try {
+        decodedToken = JSON.parse(decodedTokenString);
+        console.log("Parsed token with fields:", Object.keys(decodedToken));
+      } catch (parseError) {
+        console.error("Failed to parse token JSON:", parseError);
+        console.log("Decoded string:", decodedTokenString);
+        return createEmptyContext();
+      }
+      
+      // Validate the token (basic checks)
+      if (!decodedToken.user_id || !decodedToken.session_id) {
+        console.error("Invalid custom auth token format - missing required fields");
+        console.log("Token fields:", Object.keys(decodedToken));
+        return createEmptyContext();
+      }
+      
+      // Check token expiration
+      if (decodedToken.exp && decodedToken.exp < Math.floor(Date.now() / 1000)) {
+        console.error("Custom auth token expired");
+        return createEmptyContext();
+      }
+      
+      // Fetch the user from the database
+      const userProfile = await prisma.userProfile.findUnique({
```

Comment:
Auth tokens can never be generated on the client, as this allows an attacker to steal anyone else's session/account.

Reply:
the token gets generated on the server and get sent to the client so the signing key is still on the server.